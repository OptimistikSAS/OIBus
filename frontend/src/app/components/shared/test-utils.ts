import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };
