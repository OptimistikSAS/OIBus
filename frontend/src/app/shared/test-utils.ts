import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';
import { Page } from './types';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };

export function toPage<T>(content: Array<T>, totalElements: number = content.length, num = 0, size = Math.max(20, totalElements)): Page<T> {
  return {
    content,
    totalElements,
    number: num,
    size,
    totalPages: Math.ceil(totalElements / size)
  };
}

export function emptyPage<T>(size = 20): Page<T> {
  return {
    totalElements: 0,
    content: [] as Array<T>,
    size,
    totalPages: 0,
    number: 0
  };
}
