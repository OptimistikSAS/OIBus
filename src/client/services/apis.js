const handleResponse = async (response) => {
  try {
    const json = await response.json()
    return json
  } catch (error) {
    console.error('Error parsing JSON', error)
  }
  return null
}

const createErrorMessage = (response) => (
  `${response.status}${response.statusText ? ` - ${response.statusText}` : ''}`
)

const getRequest = async (uri) => {
  try {
    const response = await fetch(uri, { method: 'GET' })
    if (response.status !== 200) {
      throw new Error(createErrorMessage(response))
    }
    return handleResponse(response)
  } catch (error) {
    console.error('Request error', error)
    throw new Error(error)
  }
}

const downloadFileRequest = async (uri) => {
  const link = document.createElement('a')
  link.href = uri
  document.body.appendChild(link)
  link.click()
  document.body.removeChild(link)
}

const postRequest = async (uri, body) => {
  try {
    const response = await fetch(uri, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    })
    if (response.status !== 200) {
      throw new Error(createErrorMessage(response))
    }
    return response
  } catch (error) {
    throw new Error(error)
  }
}

const postTextRequest = async (uri, body) => {
  try {
    const response = await fetch(uri, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'text/plain',
      },
      body,
    })
    if (response.status !== 200) {
      throw new Error(createErrorMessage(response))
    }
    return handleResponse(response)
  } catch (error) {
    throw new Error(error)
  }
}

const putRequest = async (uri, body) => {
  try {
    const response = await fetch(uri, {
      method: 'PUT',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    })
    if (response.status !== 200) {
      throw new Error(createErrorMessage(response))
    }
    return response
  } catch (error) {
    throw new Error(error)
  }
}

const deleteRequest = async (uri) => {
  try {
    const response = await fetch(`${uri}`, {
      method: 'DELETE',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
    })
    if (response.status !== 200) {
      throw new Error(createErrorMessage(response))
    }
    return response
  } catch (error) {
    throw new Error(error)
  }
}

const getSouthProtocols = () => getRequest('/config/schemas/south')
const getNorthApis = () => getRequest('/config/schemas/north')
const getConfig = () => getRequest('config')
const getActiveConfig = () => getRequest('config/active')
const updateActiveConfig = () => putRequest('/config/activate')
const resetModifiedConfig = () => putRequest('/config/reset')
const getSouthProtocolSchema = (protocol) => getRequest(`/config/schemas/south/${protocol}`)
const getNorthApiSchema = (api) => getRequest(`/config/schemas/north/${api}`)
const addNorth = (body) => postRequest('/config/north', body)
const updateNorth = (applicationId, body) => putRequest(`/config/north/${applicationId}`, body)
const deleteNorth = (applicationId) => deleteRequest(`/config/north/${applicationId}`)
const addSouth = (body) => postRequest('/config/south', body)
const updateSouth = (dataSourceId, body) => putRequest(`/config/south/${dataSourceId}`, body)
const deleteSouth = (dataSourceId) => deleteRequest(`/config/south/${dataSourceId}`)
const updateEngine = (body) => putRequest('/config/engine', body)
const getLogs = (fromDate, toDate, verbosity) => getRequest(`logs?fromDate=${fromDate || ''}&toDate=${toDate || ''}&verbosity=[${verbosity}]`)
const getPoints = (dataSourceId) => getRequest(`/config/south/${dataSourceId}/points`)
const addPoint = (dataSourceId, body) => postRequest(`/config/south/${dataSourceId}/points`, body)
const updatePoint = (dataSourceId, pointId, body) => putRequest(`/config/south/${dataSourceId}/points/${encodeURIComponent(pointId)}`, body)
const deletePoint = (dataSourceId, pointId) => deleteRequest(`/config/south/${dataSourceId}/points/${encodeURIComponent(pointId)}`)
const deleteAllPoints = (dataSourceId) => deleteRequest(`/config/south/${dataSourceId}/points`)
const exportAllPoints = (dataSourceId) => downloadFileRequest(`/config/south/${dataSourceId}/points/export`)
const importPoints = (dataSourceId, body) => postTextRequest(`/config/south/${dataSourceId}/points/import`, body)
const getStatus = () => getRequest('status')

export default {
  getSouthProtocols,
  getNorthApis,
  getConfig,
  getActiveConfig,
  updateActiveConfig,
  resetModifiedConfig,
  getSouthProtocolSchema,
  getNorthApiSchema,
  addNorth,
  updateNorth,
  deleteNorth,
  addSouth,
  updateSouth,
  deleteSouth,
  updateEngine,
  getLogs,
  getPoints,
  addPoint,
  updatePoint,
  deletePoint,
  deleteAllPoints,
  exportAllPoints,
  importPoints,
  getStatus,
}
