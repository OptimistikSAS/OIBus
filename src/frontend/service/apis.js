const createErrorMessage = (response) => (
  `${response.status}${response.statusText ? ` - ${response.statusText}` : ''}`
)

const getRequest = async (uri) => {
  let text = ''
  try {
    const response = await fetch(uri, { method: 'GET' })
    if (response.status !== 200) {
      throw new Error(createErrorMessage(response))
    }
    text = await response.text()
    return JSON.parse(text)
  } catch (error) {
    console.error(`Request error for ${uri}:${error}`, text)
    throw new Error(error)
  }
}

const postRequest = async (uri, body) => {
  try {
    const response = await fetch(uri, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    })
    if (response.status !== 200) {
      throw new Error(createErrorMessage(response))
    }
    return JSON.parse(await response.text())
  } catch (error) {
    throw new Error(error)
  }
}

const deleteRequest = async (uri) => {
  try {
    const response = await fetch(uri, {
      method: 'DELETE',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
    })
    if (response.status !== 200) {
      throw new Error(createErrorMessage(response))
    }
    return response
  } catch (error) {
    throw new Error(error)
  }
}

const putRequest = async (uri, body) => {
  try {
    const response = await fetch(uri, {
      method: 'PUT',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    })
    if (response.status !== 200) {
      throw new Error(createErrorMessage(response))
    }
    return response
  } catch (error) {
    throw new Error(error)
  }
}

const getConfig = () => getRequest('/config')
const updateConfig = (body) => putRequest('/config', body)
const activateConfig = () => putRequest('/config/activate')

const createHistoryQuery = async (body) => postRequest('/history-queries', body)
const getHistoryQueries = async () => getRequest('/history-queries')
const getHistoryQueryById = async (id) => getRequest(`/history-queries/${id}`)
const updateHistoryQuery = async (id, body) => putRequest(`/history-queries/${id}`, body)
const enableHistoryQuery = async (id, body) => putRequest(`/history-queries/${id}/enable`, body)
const orderHistoryQuery = async (id, body) => putRequest(`/history-queries/${id}/order`, body)
const deleteHistoryQuery = async (id, position) => deleteRequest(`/history-queries/${id}?position=${position}`)
const getLastCompletedForHistoryQuery = async (id) => getRequest(`/history-queries/${id}/status`)

const getLogs = (fromDate, toDate, pageNumber, pageSize, verbosity) => getRequest(
  `/logs?fromDate=${fromDate || ''}&toDate=${toDate || ''}&pageNumber=${pageNumber}&pageSize=${pageSize}&verbosity=[${verbosity}]`,
)
const getOIBusInfo = () => getRequest('/info')
const getSouthTypes = () => getRequest('/api/installed-south')
const getNorthTypes = () => getRequest('/api/installed-north')
const getSouth = async (id) => getRequest(`/api/south/${id}`)
const getNorth = async (id) => getRequest(`/api/north/${id}`)
const getSouthStatus = (id) => getRequest(`/south/${id}/status`)
const getNorthStatus = (id) => getRequest(`/north/${id}/status`)
const reload = () => getRequest('/reload')
const shutdown = () => getRequest('/shutdown')

export default {
  getSouthTypes,
  getNorthTypes,
  getConfig,
  activateConfig,
  updateConfig,
  getHistoryQueries,
  getHistoryQueryById,
  createHistoryQuery,
  updateHistoryQuery,
  enableHistoryQuery,
  orderHistoryQuery,
  deleteHistoryQuery,
  getLastCompletedForHistoryQuery,
  getLogs,
  getOIBusInfo,
  getSouthStatus,
  getSouth,
  getNorthStatus,
  getNorth,
  reload,
  shutdown,
}
