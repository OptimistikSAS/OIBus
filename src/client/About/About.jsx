import React from 'react'
import { Container, ListGroup, ListGroupItem } from 'reactstrap'
import apis from '../services/apis'
import { AlertContext } from '../context/AlertContext.jsx'
import { ConfigContext } from '../context/ConfigContext.jsx'
import logo from '../OIBus.png'

const About = () => {
  const [staticStatus, setstaticStatus] = React.useState({})
  const [dynamicStatus, setdynamicStatus] = React.useState({})
  const { setAlert } = React.useContext(AlertContext)
  const { activeConfig } = React.useContext(ConfigContext)

  /**
   * Acquire the status
   * @returns {void}
   */
  const fetchStatus = () => {
    apis
      .getStatus()
      .then((response) => {
        setstaticStatus(response)
      })
      .catch((error) => {
        console.error(error)
        setAlert({
          text: error.message,
          type: 'danger',
        })
      })
  }

  /**
   * Fetch static status after render
   * @returns {void}
   */
  React.useEffect(() => {
    fetchStatus()
  }, [])

  /**
   * Fetch live status after render
   * @returns {void}
   */
  React.useEffect(() => {
    const source = new EventSource('/engine/sse')
    source.onerror = (error) => {
      setAlert({
        text: error.message,
        type: 'danger',
      })
    }
    source.onmessage = (event) => {
      if (event && event.data) {
        const myData = JSON.parse(event.data)
        setdynamicStatus(myData)
      }
    }
    return () => {
      source.close()
    }
  }, [activeConfig])

  return (
    <div className="p-3">
      <Container fluid>
        {staticStatus && (
          <ListGroup>
            <ListGroupItem tag="a" href="https://optimistik.io/oibus">
              <img src={logo} alt="logo" height="100px" />
            </ListGroupItem>
            <ListGroupItem>
              <b className="me-1">Version:</b>
              {staticStatus.version}
            </ListGroupItem>
            <ListGroupItem>
              <b className="me-1">Architecture:</b>
              {staticStatus.architecture}
            </ListGroupItem>
            <ListGroupItem>
              <b className="me-1">currentDirectory:</b>
              {staticStatus.currentDirectory}
            </ListGroupItem>
            <ListGroupItem>
              <b className="me-1">nodeVersion:</b>
              {staticStatus.nodeVersion}
            </ListGroupItem>
            <ListGroupItem>
              <b className="me-1">Executable:</b>
              {staticStatus.executable}
            </ListGroupItem>
            <ListGroupItem>
              <b className="me-1">ConfigurationFile:</b>
              {staticStatus.configurationFile}
            </ListGroupItem>
            <ListGroupItem>
              <b className="me-1">Hostname:</b>
              {staticStatus.hostname}
            </ListGroupItem>
            <ListGroupItem>
              <b className="me-1">Operating System:</b>
              {staticStatus.osType}
              {' '}
              {staticStatus.osRelease}
            </ListGroupItem>
            <ListGroupItem tag="a" href="https://optimistik.io/oibus">
              Official site
            </ListGroupItem>
            <ListGroupItem>
              Copyright:
              {staticStatus.copyright}
            </ListGroupItem>
            <ListGroupItem tag="a" href="https://joinup.ec.europa.eu/sites/default/files/custom-page/attachment/2020-03/EUPL-1.2%20EN.txt">
              Licensed under the EUPL-1.2-or-later
            </ListGroupItem>
            {Object.entries(dynamicStatus)
              .map(([key, value]) => (
                <ListGroupItem key={key}>
                  <div key={key}>
                    <b className="me-1">
                      {key}
                      :
                    </b>
                    <span>{value}</span>
                  </div>
                </ListGroupItem>
              ))}
          </ListGroup>
        )}
      </Container>
    </div>
  )
}
export default About
