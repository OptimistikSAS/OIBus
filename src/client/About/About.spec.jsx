/**
 * @jest-environment jsdom
 */
import React from 'react'
import ReactDOM from 'react-dom'
import { act } from 'react-dom/test-utils'

import apis from '../services/apis'
import About from './About.jsx'

import activeConfig from '../../../tests/testConfig'

let container

beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
})

// sample status (object returned by Server to give various informations on the behavior)
const status = {
  version: '0.5.7-dev',
  architecture: 'x64',
  currentDirectory: 'C:\\Users\\jfh\\Documents\\GitHub\\OIBus\\tests',
  nodeVersion: 'v12.13.1',
  executable: 'C:\\Users\\jfh\\Documents\\GitHub\\OIBus\\dist\\win\\oibus.exe',
  configurationFile: 'C:\\Users\\jfh\\Documents\\GitHub\\OIBus\\tests\\oibus.json',
  memory: '564.89/2047.61/27.59 MB/%',
  rss: '127.80/128.79/354.32 MB',
  heapTotal: '79.55/81.54/281.62 MB',
  heapUsed: '60.54/60.54/261.28 MB',
  external: '1.99/2.16/3.16 MB',
  processId: 4916,
  uptime: '3 minutes',
  hostname: 'EC2AMAZ-OBJ8M6C',
  osRelease: '10.0.14393',
  osType: 'Windows_NT',
  apisCacheStats: [
    {
      name: 'Console (points)',
      count: 195410,
      cache: 614,
    },
    {
      name: 'Console (files)',
      count: 195410,
      cache: 0,
    },
  ],
  protocolsCacheStats: [
    {
      name: 'MQTTServer',
      count: 62,
    },
    {
      name: 'SimulationServer',
      count: 422,
    },
    {
      name: 'OPC-HDA',
      count: 194926,
    },
  ],
  copyright: '(c) Copyright 2019 Optimistik, all rights reserved.',
}
// mock get Status
let resolve
let reject
const setAlert = jest.fn()
apis.getStatus = () => new Promise((_resolve, _reject) => {
  resolve = _resolve
  reject = _reject
})
React.useContext = jest.fn().mockReturnValue({ activeConfig, setAlert })
describe('About', () => {
  test('display About page based on config and status', async () => {
    act(() => {
      ReactDOM.render(
        <About />,
        container,
      )
    })
    expect(container).toMatchSnapshot()
    // resolve the call requested by useffect
    await act(async () => {
      resolve(status)
    })
    expect(container).toMatchSnapshot()
  })
  test('About: manage error in status call', async () => {
    const originalError = console.error
    console.error = jest.fn()
    act(() => {
      ReactDOM.render(
        <About />,
        container,
      )
    })
    // resolve the call requested by useffect with a reject
    await act(async () => {
      reject('error')
    })
    expect(setAlert).toHaveBeenCalled()
    console.error = originalError
  })
  test('display About with config null', () => {
    React.useContext = jest.fn().mockReturnValue({ activeConfig: null, setAlert })
    act(() => {
      ReactDOM.render(
        <About />,
        container,
      )
    })
    expect(container).toMatchSnapshot()
    React.useContext = jest.fn().mockReturnValue({ activeConfig, setAlert })
  })
})
