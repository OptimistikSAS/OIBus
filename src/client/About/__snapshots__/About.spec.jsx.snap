// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`About display About page based on config and status 1`] = `
<div>
  <div
    class="p-3"
  >
    <div
      class="container-fluid"
    >
      <ul
        class="list-group"
      >
        <a
          class="list-group-item"
          href="https://optimistik.io/oibus"
        >
          <img
            alt="logo"
            height="100px"
            src=""
          />
        </a>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            Version:
          </b>
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            Architecture:
          </b>
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            currentDirectory:
          </b>
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            nodeVersion:
          </b>
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            Executable:
          </b>
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            ConfigurationFile:
          </b>
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            Hostname:
          </b>
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            Operating System:
          </b>
           
        </li>
        <a
          class="list-group-item"
          href="https://optimistik.io/oibus"
        >
          Official site
        </a>
        <li
          class="list-group-item"
        >
          Copyright:
        </li>
        <a
          class="list-group-item"
          href="https://joinup.ec.europa.eu/sites/default/files/custom-page/attachment/2020-03/EUPL-1.2%20EN.txt"
        >
          Licensed under the EUPL-1.2-or-later
        </a>
      </ul>
    </div>
  </div>
</div>
`;

exports[`About display About page based on config and status 2`] = `
<div>
  <div
    class="p-3"
  >
    <div
      class="container-fluid"
    >
      <ul
        class="list-group"
      >
        <a
          class="list-group-item"
          href="https://optimistik.io/oibus"
        >
          <img
            alt="logo"
            height="100px"
            src=""
          />
        </a>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            Version:
          </b>
          0.5.7-dev
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            Architecture:
          </b>
          x64
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            currentDirectory:
          </b>
          C:\\Users\\jfh\\Documents\\GitHub\\OIBus\\tests
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            nodeVersion:
          </b>
          v12.13.1
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            Executable:
          </b>
          C:\\Users\\jfh\\Documents\\GitHub\\OIBus\\dist\\win\\oibus.exe
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            ConfigurationFile:
          </b>
          C:\\Users\\jfh\\Documents\\GitHub\\OIBus\\tests\\oibus.json
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            Hostname:
          </b>
          EC2AMAZ-OBJ8M6C
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            Operating System:
          </b>
          Windows_NT
           
          10.0.14393
        </li>
        <a
          class="list-group-item"
          href="https://optimistik.io/oibus"
        >
          Official site
        </a>
        <li
          class="list-group-item"
        >
          Copyright:
          (c) Copyright 2019 Optimistik, all rights reserved.
        </li>
        <a
          class="list-group-item"
          href="https://joinup.ec.europa.eu/sites/default/files/custom-page/attachment/2020-03/EUPL-1.2%20EN.txt"
        >
          Licensed under the EUPL-1.2-or-later
        </a>
      </ul>
    </div>
  </div>
</div>
`;

exports[`About display About with config null 1`] = `
<div>
  <div
    class="p-3"
  >
    <div
      class="container-fluid"
    >
      <ul
        class="list-group"
      >
        <a
          class="list-group-item"
          href="https://optimistik.io/oibus"
        >
          <img
            alt="logo"
            height="100px"
            src=""
          />
        </a>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            Version:
          </b>
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            Architecture:
          </b>
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            currentDirectory:
          </b>
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            nodeVersion:
          </b>
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            Executable:
          </b>
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            ConfigurationFile:
          </b>
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            Hostname:
          </b>
        </li>
        <li
          class="list-group-item"
        >
          <b
            class="me-1"
          >
            Operating System:
          </b>
           
        </li>
        <a
          class="list-group-item"
          href="https://optimistik.io/oibus"
        >
          Official site
        </a>
        <li
          class="list-group-item"
        >
          Copyright:
        </li>
        <a
          class="list-group-item"
          href="https://joinup.ec.europa.eu/sites/default/files/custom-page/attachment/2020-03/EUPL-1.2%20EN.txt"
        >
          Licensed under the EUPL-1.2-or-later
        </a>
      </ul>
    </div>
  </div>
</div>
`;
